crates/red_knot_python_semantic/src/types/display.rs

```diff
@@ -76,38 +101,82 @@ struct DisplayUnionType<'db> {
     db: &'db dyn Db,
 }
 
+// Enumeration for literals that are displayed in a condensed form (e.g. `Literal[... | ...]`)
+#[derive(Hash, PartialEq, Eq)]
+enum CondensedLiteral {
+    Int,
+    String,
+    Bytes,
+    Function,
+    Class,
+}
+
+impl CondensedLiteral {
+    fn iter() -> impl Iterator<Item = &'static Self> {
+        // Order of the literals in the condensed form
+        [
+            Self::Int,
+            Self::String,
+            Self::Bytes,
+            Self::Class,
+            Self::Function,
+        ]
+        .iter()
+    }
+}
+
 impl Display for DisplayUnionType<'_> {
     fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
         let union = self.ty;
 
-        let (int_literals, other_types): (Vec<Type>, Vec<Type>) = union
-            .elements(self.db)
-            .iter()
-            .copied()
-            .partition(|ty| matches!(ty, Type::IntLiteral(_)));
+        // Group literals by type: {int, string, bytes, function, class}
+        let mut literals_representation_map = HashMap::<CondensedLiteral, Vec<String>>::new();
```

Comment:
We generally try to preserve the source order of unions. Grouping them here would undo that ordering. I'm inclined that we should only group subsequent elements but I'm not sure. I'm interested to hear more opinions.

Reply:
Would you mean trying to remember the order in which we encounter the first element of each group, and condense literals in that order?

Reply:
> We generally try to preserve the source order of unions. Grouping them here would undo that ordering. I'm inclined that we should only group subsequent elements but I'm not sure. I'm interested to hear more opinions.

I would much prefer that we convert `Literal[1] | str | Literal[2]` to `Literal[1, 2] | str`. The former is much more verbose and isn't at all idiomatic (we even have lint rules against doing this if you're annotating your own code). It would be pretty clunky for us to show that in error messages, in my opinion. However, within the groups, I agree that it would be good to preserve the order as much as possible.

It's interesting that mypy and pyright take quite different approaches here. Mypy appears to always display unions as a flat list of elements, but it exactly preserves the order as the user gave them. Pyright always condenses `Literal` elements inside a union (even `Literal` elements of different types), but does not preserve the order of the union elements at all:

- https://mypy-play.net/?mypy=latest&python=3.12&gist=2a5ec4068b420e5dd94b5f4c1844bd20
- https://pyright-play.net/?pythonVersion=3.13&strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAySMApiAIYA2AsAFB0AmJwUwYYAFHVD1AB4AuQsTJUA2gEYAulAA%2BUAM4wQc4aQqUxAJikAabrzhCi68RN1QdqpSH21eUAF7GRGsQCI2YdxYAsWiwAjd0DyEHcpOgBKKABaAD4oADkwFBIBAx4QEgA3EioAfXgEEg4%2BKMyobLzC4tK4CvteavzKIsRSxyigA

Of the two I like pyright's approach more overall, but I would prefer it if we could preserve order as much as possible nonetheless

Reply:
Thank you so much for your comment and the comparison with existing implementations.

> However, within the groups, I agree that it would be good to preserve the order as much as possible
Can I understand this as:
- We do group literals by type as stated in the issue (unlike pyright)
- Within groups (e.g. ints) we show them in order of appearance
  - To be clear, this would break the previous implementation of sorting ints

> Of the two I like pyright's approach more overall, but I would prefer it if we could preserve order as much as possible nonetheless
Can I understand that as showing the groups in order of appearance?

Example:
```
Literal[1, 2] | None | Literal["A"] | Literal[0] | Literal["B"] | Unbound
    
-> Literal[1, 2, 0] | None | Literal["A", "B"] | Unbound
```
Here, we would:
- Condense by literal group
- Preserve order inside groups (order of appearance)
- Preserve order outside groups (order of appearance of the first element of the group)

That would be similar to what pyright does, except that they always place `Literal[...]` at the end and don't split them by type

All my apologies if I didn't understand correctly :)